gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/access/operations/SortedMap2IndexInsertOperation.java

```diff
@@ -192,80 +186,95 @@ else if (oldValue != null) {
+                  lockTimeout -= elapsed;
+                }
+                if (waitThreshold > 0) {
+                  // check if there is a factor of waitThreshold between
+                  // origLockTimeout and lockTimeout
+                  final long origQuotient = origLockTimeout / waitThreshold;
+                  final long quotient = lockTimeout / waitThreshold;
+                  if (origQuotient > quotient
+                      || (origLockTimeout % waitThreshold) == 0) {
+                    final LogWriterI18n logger = cache.getLoggerI18n();
+                    if (logger.warningEnabled()) {
+                      logger.warning(LocalizedStrings.LocalLock_Waiting,
+                          new Object[]{"SortedMap2IndexInsertOperation",
+                              Double.toString(waitThreshold / 1000.0),
+                              lockPolicy.getWriteLockMode().toString(),
+                              "index key with oldValue=" + ArrayUtils
+                                  .objectStringNonRecursive(oldValue) + ", owner="
+                                  + rlTXId, ArrayUtils.objectString(key),
+                              lockTimeout});
+                    }
+                    // increase waitThreshold by a factor of 2 for next iteration
+                    waitThreshold <<= 1;
                   }
-                  // increase waitThreshold by a factor of 2 for next iteration
-                  waitThreshold <<= 1;
                 }
+              } catch (InterruptedException ie) {
+                cache.getCancelCriterion().checkCancelInProgress(ie);
+              }
+              continue;
+            }
+            final ConflictException ce = new ConflictException(
+                LocalizedStrings.TX_CONFLICT_ON_OBJECT.toLocalizedString("index="
+                        + container + "; indexKey=" + ArrayUtils.objectString(key)
+                        + "; having oldValue=" + ArrayUtils.objectStringNonRecursive(
+                    oldValue) + "; owner TX=" + rlTXId
+                        + "; requested for TX=" + txId,
+                    lockPolicy.getWriteLockMode().toString()));
+            throw StandardException.newException(
+                SQLState.GFXD_OPERATION_CONFLICT, ce, ce.getMessage());
+          } else {
+            if (oldValue instanceof WrapperRowLocationForTxn) {
+              if (skipListMap.replace(key, oldValue, value)) {
+                // remove txentry, wrapper pair for reinstated map
+                // which is with TXRegionState
+                // TODO SNAP-2620 Below should be the proper way of cleanup, instead of
+                // doing a replace and then insert in GfxdTXStateProxy.cleanupIndexEntryForDestroy
+                // Check SNAP-2620 in that class.
+                // TXRegionState txrs = tx.readRegion(((GfxdTXEntryState)value).getDataRegion());
+                // txrs.getToBeReinstatedIndexMap().removeKeyPair(value, oldValue);
```

Comment:
This looks like the proper fix. Is this missing call to addUniqIdxInfosForReplacedEntries?

Reply:
Yes as commented, I wanted to do like this but this failed due to to some reason. That's why marked a todo. BTW looks like more fix is required. I extended some testing and at least one more failure is what I am seeing.